src/lsm/table_mutable.zig

```diff
@@ -42,7 +39,142 @@ pub fn TableMutableType(comptime Table: type, comptime tree_name: [:0]const u8)
         );
 
-        values: Values = .{},
+        const ValuesMap = struct {
+            const Values = struct {
+                active: [value_count_max / 8]u8,
+                items: [value_count_max]Value,
+            };
+
+            const load_factor = 50;
+            const IndexMap = std.HashMapUnmanaged(u32, void, IndexMapContext, load_factor);
+
+            const IndexMapContext = struct {
+                values: *const Values,
+                search_key: Key,
+                search_index: u32 = math.maxInt(u32),
+
+                inline fn key_from_index(ctx: *const IndexMapContext, index: u32) Key {
+                    if (index == ctx.search_index) return ctx.search_key;
+                    return key_from_value(&ctx.values.items[index]);
+                }
+
+                pub inline fn eql(ctx: IndexMapContext, a: u32, b: u32) bool {
+                    return compare_keys(ctx.key_from_index(a), ctx.key_from_index(b)) == .eq;
+                }
+
+                pub inline fn hash(ctx: IndexMapContext, index: u32) u64 {
+                    const key = ctx.key_from_index(index);
+                    return std.hash.Wyhash.hash(0, mem.asBytes(&key));
+                }
+            };
+
+            index_map: IndexMap,
+            values: *Values,
+            active: u32,
+
+            pub fn init(allocator: mem.Allocator) !ValuesMap {
+                var index_map = IndexMap{};
+                try index_map.ensureTotalCapacityContext(allocator, value_count_max, undefined);
+                errdefer index_map.deinit(allocator);
+
+                const values = try allocator.create(Values);
+                errdefer allocator.destroy(values);
+
+                return ValuesMap{
+                    .index_map = index_map,
+                    .values = values,
+                    .active = 0,
+                };
+            }
+
+            pub fn deinit(map: *ValuesMap, allocator: mem.Allocator) void {
+                allocator.destroy(map.values);
+                map.index_map.deinit(allocator);
+            }
+
+            pub inline fn count(map: *const ValuesMap) u32 {
+                return map.index_map.count();
+            }
+
+            pub inline fn clear(map: *ValuesMap) void {
+                map.index_map.clearRetainingCapacity();
+                map.active = 0;
+            }
+
+            pub fn find(map: *const ValuesMap, key: Key) ?*const Value {
+                const ctx = IndexMapContext{ .values = map.values, .search_key = key };
+                const index = map.index_map.getKeyAdapted(ctx.search_index, ctx) orelse return null;
+                return &map.values.items[index];
+            }
+
+            pub fn upsert(map: *ValuesMap, value: *const Value) *Value {
+                const ctx = IndexMapContext{ .values = map.values, .search_key = key_from_value(value) };
+                const result = map.index_map.getOrPutAssumeCapacityContext(ctx.search_index, ctx);
```

Comment:
Is this correct? Always passing `maxInt(u32)` (`ctx.search_index`) as the key to `getOrPutAssumeCapacityContext()`?

We need a fuzz test for this structure to make sure we're covering all the functionality we expect from it.

Reply:
`IndexMap` deals with indexes instead of Keys. `maxInt(u32)` is used as a dummy index to represent the key being passed in for lookup instead of the Key of a Value already in the array. `IndexMapContext` does the appropriate conversion for the hashmap.

Reply:
This was replaced with a custom map implementation.